build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavadocPlugin.java

```diff
@@ -0,0 +1,134 @@
+import java.util.List;
+
+// Handle javadoc dependencies across projects. Order matters: the linksOffline for
+// org.elasticsearch:elasticsearch must be the last one or all the links for the
+// other packages (e.g org.elasticsearch.client) will point to server rather than
+// their own artifacts.
+public class ElasticsearchJavadocPlugin implements Plugin<Project> {
+    private Project project;
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        // ignore missing javadocs
+        project.getTasks().withType(Javadoc.class).configureEach(javadoc -> {
+            // the -quiet here is because of a bug in gradle, in that adding a string option
+            // by itself is not added to the options. By adding quiet, both this option and
+            // the "value" -quiet is added, separated by a space. This is ok since the javadoc
+            // command already adds -quiet, so we are just duplicating it
+            // see https://discuss.gradle.org/t/add-custom-javadoc-option-that-does-not-take-an-argument/5959
+            javadoc.getOptions().setEncoding("UTF8");
+            ((StandardJavadocDocletOptions) javadoc.getOptions()).addStringOption("Xdoclint:all,-missing", "-quiet");
+        });
+
+        // Relying on configurations introduced by the java plugin
+        this.project.getPlugins().withType(JavaPlugin.class, javaPlugin -> project.afterEvaluate(project1 -> {
+            var withShadowPlugin = project1.getPlugins().hasPlugin(ShadowPlugin.class);
+            var compileClasspath = project.getConfigurations().getByName("compileClasspath");
+
+            if (withShadowPlugin) {
+                var shadowConfiguration = project.getConfigurations().getByName("shadow");
+                var shadowedDependencies = shadowConfiguration.getAllDependencies();
+                var nonShadowedCompileClasspath = compileClasspath.copyRecursive(
+                    dependency -> shadowedDependencies.contains(dependency) == false
+                );
+                configureJavadocForConfiguration(false, nonShadowedCompileClasspath);
+                configureJavadocForConfiguration(true, shadowConfiguration);
+            } else {
+                configureJavadocForConfiguration(false, compileClasspath);
+            }
+        }));
+    }
+
+    private void configureJavadocForConfiguration(boolean shadow, Configuration configuration) {
+        configuration.getAllDependencies()
+            .stream()
+            .sorted(Comparator.comparing(Dependency::getGroup))
+            .filter(d -> d instanceof ProjectDependency)
+            .map(d -> (ProjectDependency) d)
+            .filter(p -> p.getDependencyProject() != null)
+            .forEach(projectDependency -> configureDependency(shadow, projectDependency));
+    }
+
+    private void configureDependency(boolean shadowed, ProjectDependency dep) {
+        var upstreamProject = dep.getDependencyProject();
+        if (shadowed) {
```

Comment:
What projects are we actually publishing javadocs for that _also_ use the shadow plugin? I thought it used to only be the HLRC but that is no-more. Otherwise I think we also shadow the sql cli, but that's a CLI tool so we shouldn't be publishing Javadocs. Anything else?

Reply:
there's:

- `client/rest-high-level`
- `sql/jdbc`
- `sql/sql-cli`
- `repository-hdfs/hadoop-client-api`

from reading https://github.com/elastic/infra/blob/master/cd/release/release-manager/project-configs/master/elasticsearch.gradle I think from those only `sql/jdbc` is published to maven which includes javadoc I guess. Am I correct here?